metals/src/main/scala/scala/meta/internal/metals/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,653 @@
+package scala.meta.internal.metals
+
+import scala.meta.internal.metals.JsonParser._
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.mtags.Semanticdbs
+import scala.meta.internal.semanticdb.TextDocuments
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.{semanticdb => s}
+
+import scala.meta.internal.parsing.Trees
+import scala.meta.internal.remotels.RemoteLanguageServer
+
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.lsp4j.CallHierarchyPrepareParams
+import org.eclipse.lsp4j.CallHierarchyIncomingCallsParams
+import org.eclipse.lsp4j.CallHierarchyOutgoingCallsParams
+import org.eclipse.lsp4j.CallHierarchyIncomingCall
+import org.eclipse.lsp4j.CallHierarchyOutgoingCall
+import org.eclipse.lsp4j.CallHierarchyItem
+import org.eclipse.{lsp4j => l}
+import scala.meta.Tree
+import scala.meta.Defn
+import scala.meta.Pat
+import scala.meta.Name
+import com.google.gson.JsonElement
+import scala.meta.Member
+import scala.meta.Term
+import scala.meta.pc.CancelToken
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.Decl
+import scala.meta.Type
+import scala.meta.Init
+import scala.meta.Template
+
+final case class CallHierarchyProvider(
+    workspace: AbsolutePath,
+    semanticdbs: Semanticdbs,
+    buffers: Buffers,
+    definition: DefinitionProvider,
+    references: ReferenceProvider,
+    icons: Icons,
+    compilers: () => Compilers,
+    remote: RemoteLanguageServer,
+    trees: Trees,
+    buildTargets: BuildTargets,
+) extends SemanticdbFeatureProvider {
+
+  import CanBuildCallHierarchyItem._
+
+  override def reset(): Unit = ()
+
+  override def onDelete(file: AbsolutePath): Unit = ()
+
+  override def onChange(docs: TextDocuments, file: AbsolutePath): Unit = ()
+
+  private implicit val symbolOccurence2CanBuildCallHierarchyItem =
+    new CanBuildCallHierarchyItemBuilder[SymbolOccurrence] {
+      def apply(
+          occurence: SymbolOccurrence
+      ): CanBuildCallHierarchyItem[SymbolOccurrence] = SymbolOccurenceCHI(
+        occurence
+      )
+    }
+
+  private implicit val selectTree2CanBuildCallHierarchyItem =
+    new CanBuildCallHierarchyItemBuilder[s.SelectTree] {
+      def apply(
+          selectTree: s.SelectTree
+      ): CanBuildCallHierarchyItem[s.SelectTree] = SelectTreeCHI(selectTree)
+    }
```

Comment:
It's kind of rule of thumb to:
- insert explicit types for implicits
- define implicits in companion object of typeclass (trait) or datatype for which you want to provide them. In this case they can be moved to the companion object of `CanBuildCallHierarchyItemBuilder`


Moreover, traits and abstract classes with only one abstract method are called Single Abstract Types (SAM Types) and can be implemented as it was a function

Using that knowledge, you can do:
```scala
  object CanBuildCallHierarchyItemBuilder {
    implicit val symbolOccurence2CanBuildCallHierarchyItem: CanBuildCallHierarchyItemBuilder[SymbolOccurrence] =
      occurence => SymbolOccurenceCHI(occurence)

    implicit val selectTree2CanBuildCallHierarchyItem: CanBuildCallHierarchyItemBuilder[s.SelectTree] =
      selectTree => SelectTreeCHI(selectTree)
  }
```